### PR TITLE
[onednn] Fix qconv2d_pointwise binary ops to not alias inputs

### PR DESCRIPTION
**Pull request by frgossen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #177171

The onednn::qconv2d_pointwise.binary and .binary_tensor ops mutate
qaccum (using it as the convolution output buffer) but previously
returned it directly. This violates the PyTorch invariant that custom
operators' outputs must not alias any inputs (enforced by
check_aliasing_constraint at eager dispatch, FakeTensor tracing, and
AOT autograd runtime — becoming a hard error in PyTorch 2.12).

The old schema declared `-> Tensor(a!)` which silently bypassed the
AOT autograd check via `_schema_allows_aliasing`. Fix the
implementation to return clones at every return path in
_quantized_convolution_onednn and _fp8_convolution_onednn_ref, and
update the schemas to `Tensor(a!) qaccum -> Tensor` so the declared
contract matches the actual behavior.

Note: the is_1d + has_accum_postop_sum combination is unreachable
because run_pointwise_binary enforces act.dim() == 4.

Authored with Claude.

cc @ezyang @gchanan @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01

### CODE REVIEW

**Review by aorenste — APPROVED** [review]

**Review by mlazos — APPROVED** [review]

**Comment by zou3519** [comment]
are there torch.compile tests for this somewhere? I don't think these are going to pass them. basically the problem is torch.compile will reject custom operators that have alias annotations like these

File: aten/src/ATen/native/quantized/library.cpp [hunk 1]
```diff
@@ -265,8 +265,8 @@ TORCH_LIBRARY(onednn, m) {
...
+  m.def(TORCH_SELECTIVE_SCHEMA("onednn::qconv2d_pointwise.binary(Tensor qx, float x_scale, int x_zero_point, Tensor qw, Tensor w_scale, Tensor w_zero_point, Tensor(a!) qaccum, Tensor? bias, int[] stride, int[] padding, int[] dilation, int groups, float output_scale, int output_zero_point, ScalarType? output_dtype, float accum_scale, int accum_zero_point, str binary_attr, Scalar? alpha, str? unary_attr, Scalar?[] unary_scalars, str? unary_algorithm) -> Tensor(a)"));
+  m.def(TORCH_SELECTIVE_SCHEMA("onednn::qconv2d_pointwise.binary_tensor(Tensor qx, Tensor x_scale, Tensor x_zero_point, Tensor qw, Tensor w_scale, Tensor w_zero_point, Tensor(a!) qaccum, Tensor? bias, int[] stride, int[] padding, int[] dilation, int groups, float output_scale, int output_zero_point, ScalarType? output_dtype, float accum_scale, int accum_zero_point, str binary_attr, Scalar? alpha, str? unary_attr, Scalar?[] unary_scalars, str? unary_algorithm) -> Tensor(a)"));
```
Comment (jbschlosser) [inline aten/src/ATen/native/quantized/library.cpp]: I think the correct syntax for the return value is `Tensor(a!)` and not `Tensor(a)`, no?

**Review by jbschlosser — COMMENTED** [review]
The only precedent I'm aware of for changing op schemas is related to batch norm. The `native_batch_norm`, `cudnn_batch_norm`, and `miopen_batch_norm` ops have incorrect schemas; they mutate `running_mean`/`running_var` without declaring it via alias annotations. An attempted fix was made in the `_native_batch_norm_legit*` series of ops: https://github.com/pytorch/pytorch/blob/b442fc72198e71268fffc735b814290dda02e828/aten/src/ATen/native/native_functions.yaml#L4479-L4487

but it looks like this migration was never fully completed. Also, the batch norm situation is more complex:
* Those are ATen ops subject to strict BC / FC compatibility guarantees, including the 2-week FC window referenced above
* The `running_mean` / `running_var` Tensors that may be mutated are optional
* It looks like the custom ops in *this PR* are generally expected to be injected by inductor's quantization fusion passes rather than called directly

>  Is adding alias annotations to the schema a BC-breaking change?

I think we have to assume yes, in general. From what I can tell, the main BC risk here is that these are serialized with the old signatures through export. Loading them from a newer version of PyTorch would raise a [`SpecViolationError`](https://github.com/pytorch/pytorch/blob/main/torch/_export/verifier.py#L262-L266) since the op post-schema-change would be considered "non-functional".

A conservative approach to dealing with this could be the following:
* Define a new set of (private) ops with the correct schemas, pointing to the same implementations, with correct meta registrations
* Update inductor's quantization fusion passes to use the new ops + update AOTI shims for new ops
* Keep the old ops around for BC issues related to serialization for some length of time; remove them once we are confident serialized models will no longer contain them

I'm not too familiar with how likely it is that these ops will end up in serialized models. Also AFAICT we don't have strong BC guarantees for exported serialized models being loadable from newer PyTorch versions, so I think the risk is low.

TL;DR: I'm good with making the change directly; this PR should still be marked as BC-breaking but we can suppress the checks. will defer to @albanD though if he thinks differently :)

File: aten/src/ATen/native/quantized/library.cpp (hunk 1 above)
Comment (frgossen) [reply aten/src/ATen/native/quantized/library.cpp]: Ty. Also applied that in https://github.com/pytorch/pytorch/pull/177170/changes
